### DISCUSSION TITLE
Add the automated addition of the new version to the CHANGELOG.md file

### DISCUSSION THREAD

**Discussion by ScottWilliamAnderson** [callowayproject/bump-my-version#124] [Ideas]
bump-my-version is brilliant for automating new versions. We use it for 100% of our new versions across all our internal packages! But we're struggling to find an easy way to automatically add the version and (ideally) the short description of the release into the CHANGELOG.md file of each repository that we use it in.

What do you think of having bump-my-version include a way to automatically append the new release, and perhaps add a way to describe the changes in the newest version?

Ah, as a final note, we don't use Github for hosting the repository, so we can't make use of Github's release notes functionality unfortunately.

Thanks! Looking forward to hear what you think

**Comment by kdeldycke** [comment]
Automating the whole release workflow is doable, but complex. I managed to do it for my projects by relying on GitHub reused workflows. All my code is available here at: https://github.com/kdeldycke/workflows

There are a lot of moving parts and this is tailored to my preferences and needs. Took me basically 2 years to have something fully automated.

**Reply by ScottWilliamAnderson** [reply]
Thank you Kevin! That's a very useful repository to look into. 

I was thinking of something simpler though, like the option to enable a mandatory input prompt following a bumpversion, which would just be the short description of the release that gets appended to the release log file?

Then we'd let the usual deployment pipeline scripts take care of the rest, a bit like you do in yours :)

Best

Scott

**Comment by kdeldycke** [comment]
@coordt is also maintaining a [`generate-changelog`](https://github.com/callowayproject/generate-changelog) project. I did not take a look at it yet but might also address some of your needs.

**Comment by coordt** [comment]
@ScottWilliamAnderson  @kdeldycke I have thought about documenting the automation of this project.

It generates a version hint as part of PRs, does GitHub releases, CHANGELOGs, etc.

And boy, let me tell you, debugging an automation pipeline sucks! But this one is pretty dialed in.

What other steps would be necessary?

**Reply by ScottWilliamAnderson** [reply]
Oh that sounds wonderful! Do you already have it running for any of your projects?

In regards to what else is needed: you may have already done too much :) I was simply thinking of adding a default "add to changelog file" option, which would append the newly released (stable) versions to the changelog.md file of the repo, alongside a (ideally) required short description.

Do you already have something like this working?

Note, we'd use bump-my-version instead of GitHub automations. For this narrow of a task, I think that makes sense right?

Scott

**Reply by kdeldycke** [reply]
> And boy, let me tell you, debugging an automation pipeline sucks! But this one is pretty dialed in.

Oh this. 👆 So much this. 😖 Spent 2 years trying to automate my release workflow. 😩

I'm pretty sure there is a market for a single binary that automate all of this, like `ruff` did with linting and formatting. Or like `dependabot` did with dependency management. A highly configurable but fully integrated one-stop-shop to handle the release process (tagging, changelog update, binary compilation, distribution to package depots, version bumps, ...). 🤔

The insight would be to rely as little as possible on GitHub actions. GHA should just be an entry point. The model of their workflow is quite leaky. I was expecting something like a Petri net, but it just a loosely tied sequence of tasks. Parallel branches are hard to converge to a synchronized completion state, without workaround and hacks. So at this point, execution logic would be better encoded in that unique binary I dream of. 😅

Any volunteer? 🙃

**Reply by kdeldycke** [reply]
Two and a half years later, I ended up being the volunteer. 🙃

The single binary I was dreaming about now exists and is called [`repomatic`](https://github.com/kdeldycke/repomatic). It grew out of my reuseable workflows and turned into a Python CLI that owns all the release logic.

It still relies on `bump-my-version` for the actual bumps. But `repomatic` drives everything else: changelog updates, git tagging, GitHub releases, PyPI publishing with attestations, and Nuitka-compiled binaries for Linux, macOS and Windows.

Current limitation is some parts of `repomatic` are still GitHub-centric, but if there is demand and users, I will make it more agnostic.
